tfe/data_source_state_outputs.go

```diff
@@ -0,0 +1,260 @@
+	ValueTypeRaw json.RawMessage `json:"type"`
+}
+
+type outputData struct {
+	Value cty.Value
+}
+
+type stateData struct {
+	outputs map[string]*outputData
+}
+
+func (d dataSourceStateOutputs) readStateOutput(ctx context.Context, tfeClient *tfe.Client, orgName, wsName string) (*stateData, error) {
+	log.Printf("[DEBUG] Reading the Workspace %s in Organization %s", wsName, orgName)
+	ws, err := tfeClient.Workspaces.Read(ctx, orgName, wsName)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading workspace: %v", err)
+	}
+
+	log.Printf("[DEBUG] Reading the current StateVersion for Workspace ID %s.", ws.ID)
+	sv, err := tfeClient.StateVersions.Current(ctx, ws.ID)
+	if err != nil {
+		if err == tfe.ErrResourceNotFound {
+			return nil, fmt.Errorf("Current remote state for workspace '%s' not found.", wsName)
+		}
+		return nil, fmt.Errorf("Could not read the current state for workspace '%s' : %v", wsName, err)
+	}
+
+	log.Printf("[DEBUG] Downloading State Version")
+	rawState, err := tfeClient.StateVersions.Download(ctx, sv.DownloadURL)
+	if err != nil {
+		return nil, fmt.Errorf("Error downloading state: %v", err)
+	}
+
+	log.Printf("[DEBUG] Unmarshalling state output")
+	read := bytes.NewReader(rawState)
+	src, err := ioutil.ReadAll(read)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read state data: %v", err)
+	}
+	rrs := &rootModule{}
+	err = json.Unmarshal(src, rrs)
+	if err != nil {
+		return nil, fmt.Errorf("Could not unmarshal state data: %v", err)
+	}
+
+	fov := &stateData{
+		outputs: map[string]*outputData{},
+	}
+	for name, fos := range rrs.RootOutputs {
+		ty, err := ctyjson.UnmarshalType([]byte(fos.ValueTypeRaw))
+		if err != nil {
+			return nil, fmt.Errorf("Could not unmarshal type: %v", err)
+		}
+
+		val, err := ctyjson.Unmarshal([]byte(fos.ValueRaw), ty)
+		if err != nil {
+			return nil, fmt.Errorf("Could not unmarshal value: %v", err)
+		}
+
+		fov.outputs[name] = &outputData{
+			Value: val,
+		}
+	}
+
+	return fov, nil
+}
+
+func parseStateOutput(stateOutput *stateData) (map[string]tftypes.Value, map[string]tftypes.Type, error) {
+	tftypesValues := map[string]tftypes.Value{}
+	stateTypes := map[string]tftypes.Type{}
+
+	for name, output := range stateOutput.outputs {
+		marshData, err := output.Value.Type().MarshalJSON()
+		if err != nil {
+			return nil, nil, fmt.Errorf("Could not marshall output type: %v", err)
+		}
+		tfType, err := tftypes.ParseJSONType(marshData)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Could not parse json type data: %v", err)
+		}
+		mByte, err := ctyjson.Marshal(output.Value, output.Value.Type())
+		if err != nil {
+			return nil, nil, fmt.Errorf("Could not marshal output value and output type: %v", err)
+		}
+		tfRawState := tfprotov5.RawState{
+			JSON: mByte,
+		}
+		newVal, err := tfRawState.Unmarshal(tfType)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Could not unmarshal tftype into value: %v", err)
+		}
+		tftypesValues[name] = newVal
+		stateTypes[name] = tfType
```

Comment:
This is very clever, and something I hadn't thought of. It makes me slightly nervous, but... well, the JSON interface of types and values _is_ something that cty and tftypes need to agree on, and could be considered their compatibility interface with each other.

Just know that you're on shaky ground with the JSON parsing of the values because while they're _supposed_ to be compatible, that's not as proven as the msgpack encoding. We really only see values as JSON during state upgrades, I think, which is a less-explored part of the protocol than PlanResourceChange/ApplyResourceChange, both of which use msgpack to encode values.

Not saying what you're doing here is wrong, or that you should change it, just know you may be the first to find an edge case if one (somehow) exists. We have pretty extensive unit testing, but....